ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplication.scala

```diff
@@ -3,90 +3,205 @@
+            preExecutionDuplicateRejection(
+              commitContext,
+              transactionEntry,
+              commandDeduplicationDuration,
+              maybeDedupValue,
             )
+          } else {
+            duplicateRejection(commitContext, transactionEntry.submitterInfo)
           }
         }
-        .getOrElse(StepContinue(transactionEntry))
+      }
+
+      private def preExecutionDuplicateRejection(
+          commitContext: CommitContext,
+          transactionEntry: DamlTransactionEntrySummary,
+          commandDeduplicationDuration: Duration,
+          maybeDedupValue: Option[DamlCommandDedupValue],
+      )(implicit loggingContext: LoggingContext) = {
+        maybeDedupValue.collect {
+          case dedupValue if dedupValue.hasRecordTimeBounds => dedupValue.getRecordTimeBounds
+        } match {
+          case Some(recordTimeBounds) =>
+            val currentCommandMaxRecordTime = commitContext.maximumRecordTime.getOrElse(
+              throw Err.InternalError("Maximum record time is not set for pre-execution")
+            )
+            val maxDurationBetweenRecords = Duration.between(
+              parseInstant(recordTimeBounds.getMinRecordTime),
+              currentCommandMaxRecordTime.toInstant,
+            )
+            // We use min and max record time to determine if a command is a duplicate.
+            // These boundaries account for time skews.
+            // To guarantee that the interval between the the previous command record time and the rejection record time
+            // is bigger or equal compared to rejection deduplication duration we select the max duration between
+            // the passed deduplication duration and the maximum possible interval between the two commands.
+            val rejectionDeduplicationDuration =
+              Seq(maxDurationBetweenRecords, commandDeduplicationDuration).max
+            duplicateRejection(
+              commitContext,
+              transactionEntry.submitterInfo.toBuilder
+                .setDeduplicationDuration(
+                  buildDuration(rejectionDeduplicationDuration)
+                )
+                .build,
+            )
+          case None =>
+            duplicateRejection(commitContext, transactionEntry.submitterInfo)
+        }
+      }
+
+      private def duplicateRejection(
+          commitContext: CommitContext,
+          submitterInfo: DamlSubmitterInfo,
+      )(implicit loggingContext: LoggingContext) = {
+        rejections.reject(
+          DamlTransactionRejectionEntry.newBuilder
+            .setSubmitterInfo(submitterInfo)
+            // No duplicate rejection is a definite answer as the deduplication entry will eventually expire.
+            .setDefiniteAnswer(false)
+            .setDuplicateCommand(Duplicate.newBuilder.setDetails("")),
+          "the command is a duplicate",
+          commitContext.recordTime,
+        )
+      }
+
+      private def isTheCommandNotADuplicate(
```

Comment:
Nit: this private function is the first one being used, so I would move it a bit higher.

Reply:
Good catch. Moved it. 
Would be amazing to have this organized automatically